src/oic/utils/http_util.py

```diff
@@ -262,34 +265,110 @@ def _expiration(timeout, time_format=None):
 
 
-def cookie_signature(seed, *parts):
-    """Generates a cookie signature."""
-    sha1 = hmac.new(seed, digestmod=hashlib.sha1)
+def cookie_signature(key, *parts):
+    """Generates a cookie signature.
+
+       :param key: The HMAC key to use.
+       :type key: bytes
+       :param parts: List of parts to include in the MAC
+       :type parts: list of bytes or strings
+       :returns: hexdigest of the HMAC
+    """
+    assert isinstance(key, binary_type)
+    sha1 = hmac.new(key, digestmod=hashlib.sha1)
     for part in parts:
         if part:
-            sha1.update(part)
-    return sha1.hexdigest()
+            if isinstance(part, text_type):
+                sha1.update(part.encode('utf-8'))
+            else:
+                sha1.update(part)
+    return text_type(sha1.hexdigest())
 
 
-def make_cookie(name, load, seed, expire=0, domain="", path="", timestamp=""):
+def verify_cookie_signature(sig, key, *parts):
+    """Constant time verifier for signatures
+
+       :param sig: The signature hexdigest to check
+       :type sig: text_type
+       :param key: The HMAC key to use.
+       :type key: bytes
+       :param parts: List of parts to include in the MAC
+       :type parts: list of bytes or strings
+       :raises: `InvalidCookieSign` when the signature is wrong
+    """
+    assert isinstance(sig, text_type)
+    return hmac.compare_digest(sig, cookie_signature(key, *parts))
+
+
+def _make_hashed_key(parts, hashfunc='sha256'):
+    """
+    Construct a key via hashing the parts
+
+    If the parts do not have enough entropy of their
+    own, this doesn't help.
+
+    The size of the hash digest determines the size.
+    """
+    h = hashlib.new(hashfunc)
+    for part in parts:
+        if isinstance(part, text_type):
+            part = part.encode('utf-8')
+        if part:
+            h.update(part)
+    return h.digest()
+
+
+def make_cookie(name, load, seed, expire=0, domain="", path="", timestamp="",
+                enc_key=None):
     """
     Create and return a cookie
 
     :param name: Cookie name
     :param load: Cookie load
-    :param seed: A seed for the HMAC function
+    :param seed: A seed key for the HMAC function
     :param expire: Number of minutes before this cookie goes stale
     :param domain: The domain of the cookie
     :param path: The path specification for the cookie
     :param timestamp: A time stamp
+    :param enc_key: The key to use for cookie encryption.
     :return: A tuple to be added to headers
     """
     cookie = SimpleCookie()
     if not timestamp:
         timestamp = str(int(time.time()))
-    signature = cookie_signature(seed, load.encode("utf-8"),
-                                 timestamp.encode("utf-8"))
-    cookie[name] = "|".join([load, timestamp, signature])
+
+    bytes_load = load.encode("utf-8")
+    bytes_timestamp = timestamp.encode("utf-8")
+
+    # If we have an encryption key, we use an AEAD cipher instead of
```

Comment:
I would be worried that the comments will not match the implementation if changes come. Can you make a more higher level description of the function in the docstring?